rio_color/colorspace.pyx

```diff
@@ -0,0 +1,252 @@
+# cython: boundscheck=False
+# cython: cdivision=True
+# cython: wraparound=False
+from __future__ import division
+
+import numpy as np
+
+cimport numpy as np
+from libc.math cimport cos, sin, atan2
+cimport cython
+ctypedef np.float64_t FLOAT_t
+
+
+cdef struct st_rgb:
+    double r
+    double g
+    double b
+
+ctypedef st_rgb rgb
+
+cdef struct st_lch:
+    double l
+    double c
+    double h
+
+ctypedef st_lch lch
+
+
+def rgb_to_lch(r, g, b):
+    """Convert RGB colors to LCH
+
+    Parameters
+    ----------
+    r: float
+        0 to 1, Red
+    g: float
+        0 to 1, Blue
+    b: float
+        0 to 1, Blue
+
+    Returns
+    -------
+    L, C, H tuple (Lightness, Chroma, Hue)
+    H is in radians
+    """
+    cdef lch color
+    color = _rgb_to_lch(r, g, b)
+    return color.l, color.c, color.h
+
+
+def lch_to_rgb(l, c, h):
+    """Convert LCH colors to RGB
+
+    Parameters
+    ----------
+    L: float
+        Lightness
+    C: float
+        Chroma
+    H: float
+        Hue, radians
+
+    Returns
+    -------
+    r, g, b tuple (Red, Green and Blue)
+    scaled float 0..1
+    """
+    cdef rgb color
+    color = _lch_to_rgb(l, c, h)
+    return color.r, color.g, color.b
+
+
+cpdef np.ndarray[FLOAT_t, ndim=3] saturate_rgb(np.ndarray[FLOAT_t, ndim=3] arr, double satmult):
+    """Convert array of RGB -> LCH, adjust saturation, back to RGB
+    """
+    cdef double r, g, b
+    cdef lch c_lch
+    cdef rgb c_rgb
+
+    # Allocate an equal-sized array
+    cdef int a0 = arr.shape[0]
+    cdef int a1 = arr.shape[1]
+    cdef int a2 = arr.shape[2]
+    cdef np.ndarray[FLOAT_t, ndim=3] out = np.zeros((a0, a1, a2))
+
+    for j in range(arr.shape[2]):
+        for i in range(arr.shape[1]):
+            r = arr[0, i, j]
+            g = arr[1, i, j]
+            b = arr[2, i, j]
+
+            c_lch = _rgb_to_lch(r, g, b)
+            c_lch.c *= satmult
+            c_rgb = _lch_to_rgb(c_lch.l, c_lch.c, c_lch.h)
+
+            out[0, i, j] = c_rgb.r 
+            out[1, i, j] = c_rgb.g
+            out[2, i, j] = c_rgb.b
+
+    return out
+
+
+cpdef np.ndarray[FLOAT_t, ndim=3] arr_rgb_to_lch(np.ndarray[FLOAT_t, ndim=3] arr):
+    cdef double r, g, b
+    cdef lch color
+
+    # Allocate an equal-sized array
+    cdef int a0 = arr.shape[0]
+    cdef int a1 = arr.shape[1]
+    cdef int a2 = arr.shape[2]
+    cdef np.ndarray[FLOAT_t, ndim=3] out = np.zeros((a0, a1, a2))
```

Comment:
I couldn't find any advice or benchmarks on the best way to allocate an array of equal shape. This works but there may be more efficient ways.